chips/stm32f4xx/src/flash.rs

```diff
@@ -0,0 +1,667 @@
+        /// Read protect
+        RDP OFFSET(8) NUMBITS(8) [],
+        /// Not write protect
+        // NOTE: The length of this bit field varies with the chip model
+        nWRP OFFSET(16) NUMBITS(12) []
+    ],
+    OPTCR1 [
+        /// Not write protect
+        // NOTE: The length of this bit field varies with the chip model
+        nWRP OFFSET(16) NUMBITS(12) []
+    ]
+];
+
+// All chips models have the same FLASH_BASE
+const FLASH_BASE: StaticRef<FlashRegisters> =
+    unsafe { StaticRef::new(0x40023C00 as *const FlashRegisters) };
+
+/// Main Flash struct
+pub struct Flash {
+    registers: StaticRef<FlashRegisters>,
+}
+
+// All this hassle is caused by the fact that the following 4 chip models support 3 bit latency
+// values, while the other chips support 4 bit values
+#[cfg(not(any(
+    feature = "stm32f405",
+    feature = "stm32f415",
+    feature = "stm32f407",
+    feature = "stm32f417"
+)))]
+#[derive(Copy, Clone, PartialEq, Debug)]
+/// Enum representing all the possible values for the flash latency
+pub enum FlashLatency {
+    /// 0 wait cycles
+    Latency0,
+    /// 1 wait cycle
+    Latency1,
+    /// 2 wait cycles
+    Latency2,
+    /// 3 wait cycles
+    Latency3,
+    /// 4 wait cycles
+    Latency4,
+    /// 5 wait cycles
+    Latency5,
+    /// 6 wait cycles
+    Latency6,
+    /// 7 wait cycles
+    Latency7,
+    /// 8 wait cycles
+    Latency8,
+    /// 9 wait cycles
+    Latency9,
+    /// 10 wait cycles
+    Latency10,
+    /// 11 wait cycles
+    Latency11,
+    /// 12 wait cycles
+    Latency12,
+    /// 13 wait cycles
+    Latency13,
+    /// 14 wait cycles
+    Latency14,
+    /// 15 wait cycles
+    Latency15,
+}
+
+#[cfg(any(
+    feature = "stm32f405",
+    feature = "stm32f415",
+    feature = "stm32f407",
+    feature = "stm32f417"
+))]
+#[derive(Copy, Clone, PartialEq, Debug)]
+/// Enum representing all the possible values for the flash latency
+pub enum FlashLatency {
+    /// 0 wait cycles
+    Latency0,
+    /// 1 wait cycle
+    Latency1,
+    /// 2 wait cycles
+    Latency2,
+    /// 3 wait cycles
+    Latency3,
+    /// 4 wait cycles
+    Latency4,
+    /// 5 wait cycles
+    Latency5,
+    /// 6 wait cycles
+    Latency6,
+    /// 7 wait cycles
+    Latency7,
+}
+
+impl Flash {
+    // Flash constructor. It should be called when creating Stm32f4xxDefaultPeripherals.
+    pub(crate) fn new() -> Self {
+        Self {
+            registers: FLASH_BASE,
+        }
+    }
+
+    // The number of wait cycles depends on two factors: system clock frequency and the supply
+    // voltage. Currently, this method assumes 2.7-3.6V voltage supply (default value).
+    // TODO: Take into the account the power supply
+    //
+    // The number of wait states varies from chip to chip.
+    #[cfg(not(any(
+        feature = "stm32f410",
+        feature = "stm32f411",
+        feature = "stm32f412",
+        feature = "stm32f413",
+        feature = "stm32f423"
+    )))]
```

Comment:
It proved to be a real challenge to write a chip-agnostic flash driver. Conditional compilation isn't used in other chip crates, with the sole exception being `nrf5x` which uses conditional compilation in a trivial way. As someone with a long exposure to C and its preprocessor-based conditional compilation, this came as a natural solution for me. It makes the code a bit more fuzzy, but it reduces the amount of code duplication to minimum.

The only other possible solution that came to my mind is to create a `FlashTrait` which defines the required methods `get_latency()`, `set_latency()` and `get_number_wait_cycles_based_on_frequency()` and the associated type `FlashLatency`. Then every chip of the F4 series will implement this trait. However, I am not a big fan of this solution. First of all, it would involve code duplication, and as every code duplication, it is error-prone. Then I don't find this to be a more elegant solution then conditional compilation, nor does it make the code clearer.

What are your thoughts on this?